### ISSUE TITLE
Test case public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts fails due to incorrect parsing of concatenated string

### ISSUE THREAD

**Issue by Sidong-Wei** [grafana/grafana#26092]
**What happened**:
The unit test `public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts` fails, and the log is like this:
```
 FAIL  public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts (13.215 s)
  ● CloudWatchDatasource › When getting log groups › should return log groups as an array of strings

    expect(received).toEqual(expected) // deep equality

    - Expected  - 33
    + Received  + 33

      Array [
    -   "/aws/containerinsights/dev303-workshop/application",
    -   "/aws/containerinsights/dev303-workshop/dataplane",
    -   "/aws/containerinsights/dev303-workshop/flowlogs",
    -   "/aws/containerinsights/dev303-workshop/host",
    -   "/aws/containerinsights/dev303-workshop/performance",
    -   "/aws/containerinsights/dev303-workshop/prometheus",
    -   "/aws/containerinsights/ecommerce-sockshop/application",
    -   "/aws/containerinsights/ecommerce-sockshop/dataplane",
    -   "/aws/containerinsights/ecommerce-sockshop/host",
    -   "/aws/containerinsights/ecommerce-sockshop/performance",
    -   "/aws/containerinsights/watchdemo-perf/application",
    -   "/aws/containerinsights/watchdemo-perf/dataplane",
    -   "/aws/containerinsights/watchdemo-perf/host",
    -   "/aws/containerinsights/watchdemo-perf/performance",
    -   "/aws/containerinsights/watchdemo-perf/prometheus",
    -   "/aws/containerinsights/watchdemo-prod-us-east-1/performance",
    -   "/aws/containerinsights/watchdemo-staging/application",
    -   "/aws/containerinsights/watchdemo-staging/dataplane",
    -   "/aws/containerinsights/watchdemo-staging/host",
    -   "/aws/containerinsights/watchdemo-staging/performance",
    -   "/aws/ecs/containerinsights/bugbash-ec2/performance",
    -   "/aws/ecs/containerinsights/ecs-demoworkshop/performance",
    -   "/aws/ecs/containerinsights/ecs-workshop-dev/performance",
    -   "/aws/eks/dev303-workshop/cluster",
    -   "/aws/events/cloudtrail",
    -   "/aws/events/ecs",
    -   "/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182",
    -   "/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060",
    -   "/ecs/ecs-cwagent-daemon-service",
    -   "/ecs/ecs-demo-limitTask",
    -   "CloudTrail/DefaultLogGroup",
    -   "container-insights-prometheus-beta",
    -   "container-insights-prometheus-demo",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs/aws/containerinsights/dev303-workshop/host/aws/containerinsights/dev303-workshop/performance/aws/containerinsights/dev303-workshop/prometheus/aws/containerinsights/ecommerce-sockshop/application/aws/containerinsights/ecommerce-sockshop/dataplane/aws/containerinsights/ecommerce-sockshop/host/aws/containerinsights/ecommerce-sockshop/performance/aws/containerinsights/watchdemo-perf/application/aws/containerinsights/watchdemo-perf/dataplane/aws/containerinsights/watchdemo-perf/host/aws/containerinsights/watchdemo-perf/performance/aws/containerinsights/watchdemo-perf/prometheus/aws/containerinsights/watchdemo-prod-us-east-1/performance/aws/containerinsights/watchdemo-staging/application/aws/containerinsights/watchdemo-staging/dataplane/aws/containerinsights/watchdemo-staging/host/aws/containerinsights/watchdemo-staging/performance/aws/ecs/containerinsights/bugbash-ec2/performance/aws/ecs/containerinsights/ecs-demoworkshop/performance/aws/ecs/containerinsights/ecs-workshop-dev/performance/aws/eks/dev303-workshop/cluster/aws/events/cloudtrail/aws/events/ecs/aws/lambda/cwsyn-mycanary-fac97ded-f134-499a-9d71-4c3be1f63182/aws/lambda/cwsyn-watch-linkchecks-ef7ef273-5da2-4663-af54-d2f52d55b060/ecs/ecs-cwagent-daemon-service/ecs/ecs-demo-limitTaskCloudTrail/DefaultLogGroupcontainer-insights-prometheus-betacontainer-insights-prometheus-demo",
    +   "",
    +   "",
    +   "",
    +   "",
    +   "",
      ]

      109 |         'container-insights-prometheus-demo',
      110 |       ];
    > 111 |       expect(logGroups).toEqual(expectedLogGroups);
          |                         ^
      112 |     });
      113 |   });
      114 | 

      at public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts:111:25
      at step (node_modules/tslib/tslib.js:136:27)
      at Object.next (node_modules/tslib/tslib.js:117:57)
      at fulfilled (node_modules/tslib/tslib.js:107:62)


Test Suites: 1 failed, 475 passed, 476 total
Tests:       1 failed, 2 skipped, 3941 passed, 3944 total
Snapshots:   176 passed, 176 total
Time:        360.636 s, estimated 393 s
```
**What you expected to happen**:
The unit test should pass.

**How to reproduce it (as minimally and precisely as possible)**:
`make test-js`

**Anything else we need to know?**:
I looked into the failure a little bit. At first, I thought it was because the dataframe was not converted correctly on s390x machine, but as I manually decode the dataframe [here](https://github.com/grafana/grafana/blob/efbcbb838b4318eaf04161bc9216c74d0155efba/public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts#L64), it seems that the dataframe itself does not have any delimiter. This is a clip of the hex numbers converted from base64 string:
```
2f 61 77 73 2f 63 6f 6e 74 61 69 6e 65 72 69 6e 73 69 67 68 74 73 2f 64 65 76 33 30 33 2d 77 6f 72 6b 73 68 6f 70 2f 61 70 70 6c 69 63 61 74 69 6f 6e 2f 61 77 73 2f 63 6f 6e 74 61 69 6e 65 72 69 6e 73 69 67 68 74 73 2f 64 65 76 33 30 33 2d 77 6f 72 6b 73 68 6f 70 2f 64 61 74 61 70 6c 61 6e 65 2f 61 77 73 2f 63 6f 6e 74 61 69 6e 65 72 69 6e 73 69 67 68 74 73 2f 64 65 76 33 30 33 2d 77 6f 72 6b 73 68 6f 70 2f 66 6c 6f 77 6c 6f 67 73
```
and it could be converted to unicode
```
/aws/containerinsights/dev303-workshop/application/aws/containerinsights/dev303-workshop/dataplane/aws/containerinsights/dev303-workshop/flowlogs
```

So it seems that the conversion was correct, but I just wonder what mechanism could help to parse the concatenated string into something like:
```
["/aws/containerinsights/dev303-workshop/application", "/aws/containerinsights/dev303-workshop/dataplane", "/aws/containerinsights/dev303-workshop/flowlogs"]
```
as expected.

In other words, I am trying to understand what is the expected behavior here other than the byte conversion to make sure the log group names are parsed?

**Environment**:
- Grafana version: v7.0.5
- Data source type & version: N/A
- OS Grafana is installed on: Ubuntu 16.04
- User OS & Browser: Ubuntu 16.04 / built from source
- Grafana plugins: N/A
- Others: s390x architecture

**Comment by aocenas** [comment]
@ryantxu @kylebrandt can this be some arch dependent issue with arrow?

**Comment by Sidong-Wei** [comment]
Hi @kaydelaney, sorry for bothering you with this non-critical issue on s390x arch. I have been looking into this test case failure for a while, and I think it is likely due to the [dataframe](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts#L64) decoded difference on big-endian machines. However, since the string content is decoded correctly, I assume it may be due to the decoding of parsing information embedded, but I could not check that. Could you share some info with how you encoded the log groups array information into the dataframe so that I could re-generate it on my machine and compare them? Thanks.

**Comment by aocenas** [comment]
@ryantxu @kylebrandt I think one or the other should be better equipped to handle this than our team, not sure exactly what is happening on the backend in this context, adding this to backend platform team.

**Comment by Sidong-Wei** [comment]
Hi @aocenas , thanks a lot for the help. I tried to reach out to Kay because he/she was the one who added this particular test case, and I am sorry about any inconvenience this might have caused. Since this was the only test case failure in the unit tests, I hope this should be not too difficult for you to identify the problem. I will also continue investigating this, thank you again.

**Comment by Sidong-Wei** [comment]
I have solved this issue by modifying the "offset" bytes in the dataframe. It turns out that the few bytes in the hex code right before the string data are the offset bytes that controls the parsing of the array. They are 4 bytes-long and are generated on little-endian machines, so they need to be flipped on big-endian machines. To be specific, the following hex number on liitle-endian machines:
```
32 00 00 00 62 00 00 00 91 00 00 00 bc 00 00 00 ee 00 00 00 1f 01 00 00 54 01 00 00 87 01 00 00 b5 01 00 00 ea 01 00 00 1b 02 00 00 4a 02 00 00 74 02 00 00 a5 02 00 00 d5 02 00 00 10 03 00 00 44 03 00 00 76 03 00 00 a3 03 00 00 d7 03 00 00 09 04 00 00 40 04 00 00 77 04 00 00 97 04 00 00 ad 04 00 00 bc 04 00 00 fb 04 00 00 42 05 00 00 61 05 00 00 78 05 00 00 92 05 00 00 b4 05 00 00 d6 05 00 00
```
needs to be flipped into
```
00 00 00 32 00 00 00 62 00 00 00 91 00 00 00 bc 00 00 00 ee 00 00 01 1f 00 00 01 54 00 00 01 87 00 00 01 b5 00 00 01 ea 00 00 02 1b 00 00 02 4a 00 00 02 74 00 00 02 a5 00 00 02 d5 00 00 03 10 00 00 03 44 00 00 03 76 00 00 03 a3 00 00 03 d7 00 00 04 09 00 00 04 40 00 00 04 77 00 00 04 97 00 00 04 ad 00 00 04 bc 00 00 04 fb 00 00 05 42 00 00 05 61 00 00 05 78 00 00 05 92 00 00 05 b4 00 00 05 d6
```
on big-endian machines so that they will be read correctly.
After this conversion, the dataframe could be parsed correctly on s390x machines and the test will pass. Close the issue now.
